### PR TITLE
[C++ API] Document BatchNorm and update default behavior

### PR DESCRIPTION
**Pull request by goldsborough**
This PR:

1. Documents `BatchNorm`,
2. Makes a number of API changes after reconsidering some quirks:
    1. The default value for the `stateful` parameter used to be `false`, but the most common usage of `BatchNorm` out of the wild is certainly stateful, and the default in Python is also statefulness. So we change the default to stateful.
    2. The `pure_forward` function used to use the internal running mean and variance variables instead of the ones supplied to that function call when `stateful` was true, which certainly seems odd. When you call `pure_forward` you would certainly expect the values you pass explicitly to be used. This is now fixed.
3. Adds tests for `BatchNorm`, finally.

@ebetica @apaszke @ezyang

### CODE REVIEW

**Comment by SsnL** [comment]
Now do we want the `momentum` parameter to be forever wrongly named? Or do we want to make it do the correct thing in c++ api?

**Comment by goldsborough** [comment]
@SsnL I'm open to suggestions. I think it would be ok to give it a different name, especially because I don't imagine people changing it around too much typically. Maybe `decay`? Or `gamma` (like in `ExponentialLR`? 
CC @ebetica

**Comment by SsnL** [comment]
`gamma` has a special meaning in BN, but `decay` sounds reasonable. fwiw, cudnn names it `exponentialAverageFactor`.

**Review by ebetica — APPROVED** [review]

**Comment by ebetica** [comment]
I don't have a good opinion about the momentum parameter, I might defer this to some of the vision folks?

**Comment by ebetica** [comment]
@prigoyal said that momentum is a universally well understood term so renaming it here probably doesn't make too much sense.

**Review by ezyang — APPROVED** [review]
Not sure why you need code owner review on this one.

**Review by soumith — DISMISSED** [review]
Do not diverge the PyTorch and C++ APIs. If we have BatchNorm momentum at the python level, for better or worse we stick to it.
API divergence is a much worse experience than academic correctness.
